### PR TITLE
KAFKA-5093: Avoid loading full batch data when possible when iterating FileRecords

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

**Comment by hachikuji** [comment]
@ijuma or @junrao If you have time. Don't be scared by the number of lines because the changes are straightforward and a large portion of it is testing.

File: core/src/main/scala/kafka/message/MessageAndOffset.scala [hunk 1]
```diff
@@ -17,11 +17,23 @@
...
+import org.apache.kafka.common.record.{AbstractLegacyRecordBatch, Record, RecordBatch}
 
 object MessageAndOffset {
-  def fromRecordBatch(recordBatch: AbstractLegacyRecordBatch): MessageAndOffset = {
-    MessageAndOffset(Message.fromRecord(recordBatch.outerRecord), recordBatch.lastOffset)
+  def fromRecordBatch(batch: RecordBatch): MessageAndOffset = {
+    if (!batch.isInstanceOf[AbstractLegacyRecordBatch])
+      throw new IllegalArgumentException(s"Illegal batch type ${batch.getClass}. The older message format classes " +
+        s"only support conversion from ${classOf[AbstractLegacyRecordBatch]}, which is used for magic v0 and v1")
+    val legacyRecordBatch = batch.asInstanceOf[AbstractLegacyRecordBatch]
```
Comment (ijuma) [inline core/src/main/scala/kafka/message/MessageAndOffset.scala]: We should use `pattern matching` since it does the cast for you automatically.

File: core/src/main/scala/kafka/message/MessageAndOffset.scala [hunk 2]
```diff
@@ -17,11 +17,23 @@
...
+    MessageAndOffset(Message.fromRecord(legacyRecordBatch.outerRecord), legacyRecordBatch.lastOffset)
+  }
+
+  def fromRecord(record: Record): MessageAndOffset = {
+    if (!record.isInstanceOf[AbstractLegacyRecordBatch])
+      throw new IllegalArgumentException(s"Illegal record type ${record.getClass}. The older message format classes " +
+        s"only support conversion from ${classOf[AbstractLegacyRecordBatch]}, which is used for magic v0 and v1")
+    val legacyRecordBatch = record.asInstanceOf[AbstractLegacyRecordBatch]
```
Comment (ijuma) [inline core/src/main/scala/kafka/message/MessageAndOffset.scala]: Here too.

File: clients/src/test/java/org/apache/kafka/common/record/FileLogInputStreamTest.java [hunk 3]
```diff
@@ -16,47 +16,249 @@
...
+            assertEquals(records[i].key(), batchRecords.get(i).key());
+            assertEquals(records[i].value(), batchRecords.get(i).value());
+            if (magic == MAGIC_VALUE_V0)
+                assertEquals(NO_TIMESTAMP, batchRecords.get(i).timestamp());
+            else
+                assertEquals(records[i].timestamp(), batchRecords.get(i).timestamp());
+        }
+    }
+
+    @Parameterized.Parameters
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/common/record/FileLogInputStreamTest.java]: We should add parameter names here.

**Review by ijuma — COMMENTED** [review]
Left a few very quick comments. Will take a deeper look later if @junrao doesn't beat me to it.

File: clients/src/main/java/org/apache/kafka/common/record/FileLogInputStream.java [hunk 4]
```diff
@@ -265,22 +182,37 @@ public void writeTo(ByteBuffer buffer) {
...
-        @Override
-        public boolean isControlBatch() {
-            loadUnderlyingRecordBatch();
-            return underlying.isControlBatch();
+        protected RecordBatch loadBatchHeader() {
+            if (fullBatch != null)
+                return fullBatch;
+
+            if (headerOnlyBatch == null)
+                headerOnlyBatch = loadBatchWithSize(headerSize(), "record batch header");
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/record/FileLogInputStream.java]: Nit: I was thinking that the variable name should be `batchHeader` and this string seems to confirm that.

File: clients/src/main/java/org/apache/kafka/common/record/LegacyRecord.java [hunk 5]
```diff
@@ -553,6 +554,12 @@ public static int recordOverhead(byte magic) {
         return RECORD_OVERHEAD_V1;
     }
 
+    public static int headerSize(byte magic) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/record/LegacyRecord.java]: Since this is a public method, we should probably validate that `magic` is not > 1. Same for `recordOverhead` and `keyOffset`.

File: clients/src/main/java/org/apache/kafka/common/record/FileLogInputStream.java [hunk 6]
```diff
@@ -35,7 +41,7 @@
     private int position;
     private final int end;
     private final FileChannel channel;
-    private final ByteBuffer logHeaderBuffer = ByteBuffer.allocate(LOG_OVERHEAD);
+    private final ByteBuffer logHeaderBuffer = ByteBuffer.allocate(LOG_OVERHEAD_UPTO_MAGIC);
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/record/FileLogInputStream.java]: Would this be better named as `HEADER_UP_TO_MAGIC` or something like that?

File: clients/src/main/java/org/apache/kafka/common/record/FileLogInputStream.java [hunk 7]
```diff
@@ -309,7 +241,9 @@ public int hashCode() {
 
         @Override
         public String toString() {
-            return "FileChannelRecordBatch(magic: " + magic() + ", offsets: [" + baseOffset() + ", " + lastOffset() + "])";
+            return "FileChannelRecordBatch(magic: " + magic +
+                    ", offset: " + offset +
+                    ", size: " + batchSize + ")";
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/record/FileLogInputStream.java:246]: Is it worth overriding in the subclass so that we get the range of offsets?

File: clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java [hunk 8]
```diff
@@ -506,6 +506,13 @@ public static MemoryRecords withTransactionalRecords(CompressionType compression
...
     }
 
+    public static MemoryRecords withTransactionalRecords(byte magic, long initialOffset, CompressionType compressionType,
+                                                         long producerId, short producerEpoch, int baseSequence,
+                                                         int partitionLeaderEpoch, SimpleRecord... records) {
+        return withRecords(magic, initialOffset, compressionType, TimestampType.CREATE_TIME, producerId, producerEpoch,
+                baseSequence, partitionLeaderEpoch, true, records);
+    }
+
     public static MemoryRecords withTransactionalRecords(long initialOffset, CompressionType compressionType, long producerId,
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java:516]: Nit: maybe this can call the newly introduced `withTransactionalRecords`.

File: clients/src/test/java/org/apache/kafka/common/record/FileLogInputStreamTest.java [hunk 9]
```diff
@@ -16,47 +16,249 @@
...
+        this.magic = magic;
+        this.compression = compression;
+    }
+
     @Test
     public void testWriteTo() throws IOException {
         try (FileRecords fileRecords = FileRecords.open(tempFile())) {
-            fileRecords.append(MemoryRecords.withRecords(CompressionType.NONE, new SimpleRecord("foo".getBytes()),
-                    new SimpleRecord("bar".getBytes())));
+            fileRecords.append(MemoryRecords.withRecords(magic, compression, new SimpleRecord("foo".getBytes())));
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/common/record/FileLogInputStreamTest.java:60]: Why did you remove the second record?

File: clients/src/test/java/org/apache/kafka/common/record/FileLogInputStreamTest.java [hunk 10]
```diff
@@ -16,47 +16,249 @@
...
+
+        if (magic == MAGIC_VALUE_V0)
+            assertEquals(NO_TIMESTAMP_TYPE, batch.timestampType());
+        else
+            assertEquals(CREATE_TIME, batch.timestampType());
+        assertEquals(baseOffset + records.length - 1, batch.lastOffset());
+        if (magic >= MAGIC_VALUE_V2)
+            assertEquals(Integer.valueOf(records.length), batch.countOrNull());
+
+        assertEquals(baseOffset, batch.baseOffset());
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/common/record/FileLogInputStreamTest.java:244]: Should we check `maxTimestamp` and `baseTimestamp`?

File: core/src/main/scala/kafka/message/MessageAndOffset.scala [hunk 11]
```diff
@@ -17,11 +17,31 @@
...
-import org.apache.kafka.common.record.AbstractLegacyRecordBatch
+import org.apache.kafka.common.record.{AbstractLegacyRecordBatch, Record, RecordBatch}
 
 object MessageAndOffset {
-  def fromRecordBatch(recordBatch: AbstractLegacyRecordBatch): MessageAndOffset = {
-    MessageAndOffset(Message.fromRecord(recordBatch.outerRecord), recordBatch.lastOffset)
+  def fromRecordBatch(batch: RecordBatch): MessageAndOffset = {
+    batch match {
+      case legacyBatch: AbstractLegacyRecordBatch =>
+        val legacyRecordBatch = batch.asInstanceOf[AbstractLegacyRecordBatch]
```
Comment (ijuma) [inline core/src/main/scala/kafka/message/MessageAndOffset.scala]: This can be removed.

File: core/src/main/scala/kafka/message/MessageAndOffset.scala [hunk 12]
```diff
@@ -17,11 +17,31 @@
...
+      case _ =>
+        throw new IllegalArgumentException(s"Illegal batch type ${batch.getClass}. The older message format classes " +
+          s"only support conversion from ${classOf[AbstractLegacyRecordBatch]}, which is used for magic v0 and v1")
+    }
+  }
+
+  def fromRecord(record: Record): MessageAndOffset = {
+    record match {
+      case legacyBatch: AbstractLegacyRecordBatch =>
+        val legacyRecordBatch = record.asInstanceOf[AbstractLegacyRecordBatch]
```
Comment (ijuma) [inline core/src/main/scala/kafka/message/MessageAndOffset.scala]: This can be removed.

File: core/src/main/scala/kafka/message/MessageAndOffset.scala [hunk 13]
```diff
@@ -17,11 +17,31 @@
...
+
+      case _ =>
+        throw new IllegalArgumentException(s"Illegal batch type ${batch.getClass}. The older message format classes " +
+          s"only support conversion from ${classOf[AbstractLegacyRecordBatch]}, which is used for magic v0 and v1")
+    }
+  }
+
+  def fromRecord(record: Record): MessageAndOffset = {
+    record match {
+      case legacyBatch: AbstractLegacyRecordBatch =>
```
Comment (ijuma) [inline core/src/main/scala/kafka/message/MessageAndOffset.scala:36]: It's so weird how the record and record batch are of the same type. Happy that we're moving away from that.

**Review by ijuma — COMMENTED** [review]
Thanks for the PR, looks good overall, just a few minor comments.

File: clients/src/main/java/org/apache/kafka/common/record/FileLogInputStream.java (hunk 6 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/common/record/FileLogInputStream.java]: Works for me.

File: clients/src/main/java/org/apache/kafka/common/record/FileLogInputStream.java (hunk 7 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/common/record/FileLogInputStream.java:246]: Couldn't hurt.

Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/common/record/FileLogInputStream.java:246]: Hmm.. On second thought, I was trying to avoid any incidental loading from `toString`.

File: clients/src/test/java/org/apache/kafka/common/record/FileLogInputStreamTest.java (hunk 9 above)
Comment (hachikuji) [reply clients/src/test/java/org/apache/kafka/common/record/FileLogInputStreamTest.java:60]: It was so that the batch behavior would be consistent across all magic versions and compression types since the number of records in the batch didn't seem too important for this test case.

File: clients/src/main/java/org/apache/kafka/common/record/FileLogInputStream.java (hunk 7 above)
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/record/FileLogInputStream.java:246]: Makes sense.

File: clients/src/main/java/org/apache/kafka/common/record/LegacyRecord.java (hunk 5 above)
Comment (hachikuji) [reply clients/src/main/java/org/apache/kafka/common/record/LegacyRecord.java]: Found a bug in a test case after adding these checks!

**Review by ijuma — APPROVED** [review]
Thanks, LGTM.
